packages/core/src/request.ts

```diff
@@ -161,38 +167,36 @@ export class Request {
         this.retryCount = retryCount;
         this.errorMessages = [...errorMessages];
         this.headers = { ...headers };
-        this.userData = { ...userData };
+        this.#userData = { ...userData };
         this.handledAt = handledAt as unknown instanceof Date ? (handledAt as Date).toISOString() : handledAt!;
 
-        Object.defineProperties(this.userData, {
-            __crawlee: {
-                // eslint-disable-next-line no-underscore-dangle
-                value: this.userData.__crawlee,
-                enumerable: false,
-                writable: true,
-            },
-            toJSON: {
-                value: () => ({
-                    ...this.userData,
-                    // eslint-disable-next-line no-underscore-dangle
-                    __crawlee: this.userData?.__crawlee,
+        // eslint-disable-next-line no-underscore-dangle
+        this.#internalVariables = userData.__crawlee ?? {};
+
+        if (skipNavigation != null) this.skipNavigation = skipNavigation;
+
+        Object.defineProperty(this, 'userData', {
+            get: () => ({
```

Comment:
this feels also weird and slow, reading of the property will be much more common than writing, we need to optimize for reading, so we can't create fresh objects every time we read.